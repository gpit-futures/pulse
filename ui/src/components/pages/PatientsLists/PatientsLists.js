import React, { PureComponent } from 'react';
import { Row, Col } from 'react-bootstrap';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { compose, lifecycle } from 'recompose';

import PatientsList from '../../containers/PatientsList/PatientsList';
import patientsSelector from './selectors';
import { fetchPatientsRequest } from '../../../ducks/feth-patients.duck';
import { fetchPatientCountsRequest } from '../../../ducks/fetch-patient-counts.duck'
import { fetchPatientsOnMount, fetchPatientsCountsOnMountAndUpdate } from '../../../utils/HOCs/fetch-patients.utils';

const mapDispatchToProps = dispatch => ({ actions: bindActionCreators({ fetchPatientsRequest, fetchPatientCountsRequest }, dispatch) });

@connect(patientsSelector, mapDispatchToProps)
@compose(lifecycle(fetchPatientsOnMount), lifecycle(fetchPatientsCountsOnMountAndUpdate))
class PatientsLists extends PureComponent {
  render() {
    return (<section className="page-wrapper">
      <Row>
        <Col xs={12}>
          <PatientsList {...this.props} />
        </Col>
      </Row>
    </section>)
  }
}

export default PatientsLists;
