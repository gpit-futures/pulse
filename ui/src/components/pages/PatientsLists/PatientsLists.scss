.table thead tr th,
.table tbody tr td {
  &.patient-control {
    &:before { display: none; }
    padding: 5px 4px 5px 7px;
    @include clearfix;
    @include respond-to(xs) { display: none; }
  }
}
.patient-buttons {
  @include clearfix;
  float: right;
  width: 80px;
  @include respond-from(xs) {
    float: none;
    margin: 0 auto;
  }
  .btn { border-radius: 0; }
  .btn-view-patient {
    padding-left: 0;
    padding-right: 0;
    text-align: center;
    width: 100%;
  }
  .btn-dropdown-toggle {
    float: right;
    margin-left: -1px;
    min-width: auto;
    padding: 0;
    width: 20px;
    &:after {
      margin-right: 1px;
      right: 50%;
      transform: translate(50%, -50%);
    }
  }
  &.dropdown.open {
    .btn-dropdown-toggle,
    .btn-dropdown-toggle:hover,
    .btn-dropdown-toggle:focus {
      background: $base-color !important;
      &:after {
        color: $text-on-base-color
      }
    }
  }
}
@include respond-to(xs) {
  .dropdown-menu.dropdown-menu-patients {
    right: -35px;
    width: 288px;
    .btn-smaller {
      height: 30px;
      line-height: 20px;
    }
  }
}

.wrap-patients-table {
  max-width: initial;
  overflow-x: auto;
  width: 100%;
  .table {
    max-width: none;
    min-width: 100%;
    thead tr th,
    tbody tr td {
      @include respond-from(xs) { min-height: 40px; }
      word-wrap: normal;
    }
  }
}

.wrap-patients-table .table-patients-name,
.wrap-patients-table .table-patients-controls {
  min-width: 50px;
  position: absolute;
  top: 0;
  z-index: 2;
  @include respond-to(xs) { display: none; }
}
.wrap-patients-table .table-patients-name {
  border-right: 0;
  left: 0;
  width: auto;
  thead tr th:first-child,
  tbody tr td:first-child {
    border-right: 1px solid #B4B5B4;
  }
}
.wrap-patients-table .table-patients-controls {
  right: 0;
  width: auto;
}