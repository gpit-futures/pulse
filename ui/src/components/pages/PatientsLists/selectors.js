import _ from 'lodash/fp';
import qs from 'qs';
import { createSelector } from 'reselect';

import { patientsDepartments, patientsAgeRanges, AGE_RANGE_PREFIX, DEPARTMENT_PREFIX } from '../../../config/patients.constants';

const getFilterObj = (search) => {
  const query = qs.parse(search.replace('?', ''));
  const prefix = _.flow(_.keys, _.head)(query);
  const filterCriteriaLabel = _.flow(_.values, _.head)(query);

  if (prefix && filterCriteriaLabel) {
    switch (prefix) {
      case AGE_RANGE_PREFIX:
        return _.find(ageRange => ageRange.name === filterCriteriaLabel)(patientsAgeRanges)
      case DEPARTMENT_PREFIX:
        return _.find(department => department.name === filterCriteriaLabel)(patientsDepartments)
      default:
        return null;
    }
  }
  return null
};

const getFilterPredicate = _.flow(getFilterObj, _.getOr(_.stubTrue, 'predicate'));

const allPatientsSelector = ({ patients }, { location: { search } }) => {
  const allPatients = _.values(patients);

  const filterPredicate = getFilterPredicate(search);
  return _.filter(filterPredicate)(allPatients);
};

const patientsCountsSelector = ({ patientsCounts }) => patientsCounts;

const panelTitleSelector = (state, { location: { search } }) => {
  const filterObj = getFilterObj(search);
  const name = _.get('name')(filterObj);
  const prefixLabel = _.get('prefixLabel')(filterObj);

  if (name && prefixLabel) return `Patient Info. ${prefixLabel}: ${name}`;
  return 'Patient Info'
};

const patientsSelector = createSelector(
  allPatientsSelector,
  patientsCountsSelector,
  panelTitleSelector,
  (allPatients, patientsCounts, panelTitle) => {
    const getPatientCounts = id =>_.getOr({}, [id, 0], patientsCounts); //TODO maybe patient counts can be not only [{}] but [{}, ..., {}]
    const allPatientsWithCounts = _.map(({ id, ...restPatient }) => ({ id, ...restPatient, ...getPatientCounts(id) }), allPatients);
    return ({ allPatients, patientsCounts, allPatientsWithCounts, panelTitle });
  }
);

export default patientsSelector;
