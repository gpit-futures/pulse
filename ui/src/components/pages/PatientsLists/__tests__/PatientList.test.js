import React from 'react';
import Enzyme, { shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';
import configureStore from 'redux-mock-store';

import PatientsLists from '../PatientsLists';

Enzyme.configure({ adapter: new Adapter() });

const mockStore = configureStore();
const store = mockStore();
const match = {
	params: {},
};
const location = {
	pathname: '/patients',
		search: '?ageRange=61-80',
};
const context = {
  router: {
    history: {
      push: () => {},
      replace: () => {},
      location: {
        pathname: '/patients',
        search: '?ageRange=61-80',
      },
    },
    route: {
      match: {
        params: {},
      },
    },
  },
};

describe('Component <PatientsLists />', () => {
  it('should renders correctly', () => {
    const component = shallow(
      <PatientsLists
        store={store}
        match={match}
        location={location}
      />, { context }).dive().dive().dive();

    expect(component).toMatchSnapshot();

    expect(component.find('.page-wrapper')).toHaveLength(1);
    expect(component.find('PatientsList')).toHaveLength(1);
  });
});

