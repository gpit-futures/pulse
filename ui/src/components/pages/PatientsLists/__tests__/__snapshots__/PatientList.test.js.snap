// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component <PatientsLists /> should renders correctly 1`] = `
<section
  className="page-wrapper"
>
  <Row
    bsClass="row"
    componentClass="div"
  >
    <Col
      bsClass="col"
      componentClass="div"
      xs={12}
    >
      <PatientsList
        actions={
          Object {
            "fetchPatientCountsRequest": [Function],
            "fetchPatientsRequest": [Function],
          }
        }
        allPatients={Array []}
        allPatientsWithCounts={Array []}
        location={
          Object {
            "pathname": "/patients",
            "search": "?ageRange=61-80",
          }
        }
        match={
          Object {
            "params": Object {},
          }
        }
        panelTitle="Patient Info. Age range: 61-80"
        patientsPerPageAmount={10}
        store={
          Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
        }
        storeSubscription={
          Subscription {
            "listeners": Object {
              "clear": [Function],
              "get": [Function],
              "notify": [Function],
              "subscribe": [Function],
            },
            "onStateChange": [Function],
            "parentSub": undefined,
            "store": Object {
              "clearActions": [Function],
              "dispatch": [Function],
              "getActions": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
            },
            "unsubscribe": [Function],
          }
        }
      />
    </Col>
  </Row>
</section>
`;
